elastic/beats
# [Metricbeat] Add latency config option into aws module

## What does this PR do?

This PR adds the `latency` config parameter for `aws` module. With this config parameter, users can collect CloudWatch metrics with a latency larger than Metricbeat collection period. For example, for S3 request metrics, the latency can be 3 minutes. The recommended collection period is 1min. Without the `latency` config parameter, no data points will be collected ever. 

In this screenshot: current timestamp is 14:21 but we only see the latest data at 14:18 with a 3-minute delay. In this case, without this fix, our Metricbeat module will keep trying to collect data from the last 2-minute time frame, which will always be empty:
<img width="499" alt="Screenshot 2020-09-03 at 16 22 19" src="https://user-images.githubusercontent.com/14081635/92127724-d5842d80-edbe-11ea-842a-a71210bc4df2.png">

(@gbanasiak Thank you for your screenshot!)

## Why is it important?

`latency` config sets CloudWatch API start time and end time with a delay. For example, if the current time is `2020-08-31T18:30:00.000Z` with collection period `1m`. 

Without `latency`:
startTime = `2020-08-31T18:29:00.000Z`
endTime = `2020-08-31T18:30:00.000Z`

With `latency` set to `5min`:
startTime = `2020-08-31T18:24:00.000Z`
endTime = `2020-08-31T18:25:00.000Z`

## Checklist




## How to test this PR locally
1. Enable aws module: `./metricbeat modules enable aws`
2. Change `modules.d/aws.yml` to:
```
- module: aws
  period: 5m
  credential_profile_name: elastic-beats
  latency: 5m
  metricsets:
    - ec2
```
3. Start Metricbeat with debug level logging: `./metricbeat -e -d "*"`
4. You should see the startTime and endTime in debug log with a latency of 5 minutes applied.
